spec/app/ics-029-fee-payment/README.md

```diff
@@ -96,7 +99,33 @@ function PayTimeoutFee(packet: Packet, timeout_relayer: string) {
 }
 ```
 
-These functions will then be utilized in the ICS-4 handlers like so:
+
+The fee module should also expose the following queries so that relayers may query their expected fee:
+
+```typescript
+// Gets the fee expected for submitting ReceivePacket msg for this packet
+function GetReceiveFee(packet) Fee
+
+// Gets the fee expected for submitting AcknowledgePacket msg for this packet
+function GetAckFee(packet) Fee
+
+// Gets the fee expected for submitting TimeoutPacket msg for this packet
+function GetTimeoutFee(packet) Fee
+```
+
+Since different chains may have different representations for fungible tokens and this information is not being sent to other chains; this ICS does not specify a particular representation for the `Fee`. Each chain may choose its own representation, it is incumbent on relayers to interpret the Fee correctly.
+
+### Connection Negotiation
+
+The chains must agree to enable the incentivization feature during the connection handshake. This can be done by adding an incentivization feature to the connection version.
+
+```{"1", ["ORDER_ORDERED", "ORDER_UNORDERED", "INCENTIVE_V1"]}```
```

Comment:
I like the `"INCENTIVE_V1"` idea. Adding features to the connection.

However, is this opt-in per channel? How does it work?

`ORDER_ORDERED` means "we can create an ordered channel", but doesn't change the functionality for those not using an ordered channel.

Do we define "ordered, incentivized channel"? Or is this orthogonal?

Reply:
Good question. My thinking is that all channels (whether they are ORDERED or UNORDERED) built on top of a connection with `INCENTIVE_V1` are incentivized channels. The ICS-4 handlers check if this feature is enabled in their underlying connection and do the appropriate logic.
All channels built on top of connection without that feature are unincentivized channels.

I think we want to enable two cases:
- Ability for an incentivized IBC to "downgrade" and connect with an unincentivized chain
- Ability for a module to enforce that it is connected by incentivized channel.

I'm not sure if there's a use for opting to not have incentivization when it is possible. Especially if the ICS-4 handlers are dealing with all fee logic, why would it matter to an app that its packets are getting incentivized?
In the case where fees are unnecessary for an app, users can just set 0 fees. There's no reason the app has to enforce this. Is there a use case you had in mind?

I think handling it this way is the easiest way to not require applications to bump their versions as a way of signalling that their underlying chain supports incentivization.

I'd prefer not to create channel types that grow in size on the order of the cross-product of available features.

An ideal solution I think would be to create a feature at the connection level, and then allow channels to opt-in to their desired subset of features.

This doesn't seem to be enabled by the channel handshake. It would be nice if the channel handshake allowed relayer to specify a subset of connection features it would like to use, but this is not possible since ORDER is the only specific choice provided to relayers on channel handshake. We should look into changing this for future versions of IBC.

However, given the lack of a use case for apps voluntarily barring incentivization and the desire to not communicate a core IBC feature in the channel version which is controlled by apps; I think this makes the most sense as an approach. However, I'm open to better ideas here.

Reply:
> I'm not sure if there's a use for opting to not have incentivization when it is possible. Especially if the ICS-4 handlers are dealing with all fee logic, why would it matter to an app that its packets are getting incentivized?
In the case where fees are unnecessary for an app, users can just set 0 fees. There's no reason the app has to enforce this. Is there a use case you had in mind?
>
> I think handling it this way is the easiest way to not require applications to bump their versions as a way of signalling that their underlying chain supports incentivization.
>
> I'd prefer not to create channel types that grow in size on the order of the cross-product of available features.

I agree with you here.

Reply:
My question was since a channel on a connection with ```{"1", ["ORDER_ORDERED", "ORDER_UNORDERED", "INCENTIVE_V1"]}``` is incompatible with a channel on connection with ```{"1", ["ORDER_ORDERED", "ORDER_UNORDERED"]}```... why not just bump the version and call this

```{"2", ["ORDER_ORDERED", "ORDER_UNORDERED"]}```

Reply:
Ahh yes, this would work.

In that case, a chain that supports incentivization would send over possible connection versions: `{{"1", ["ORDER_ORDERED", "ORDER_UNORDERED"]}, {"2", ["ORDER_ORDERED", "ORDER_UNORDERED"]}}`

And the counterparty chain will pick the highest that it can support.